pkg/notifications/notification.go

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2024 Greenbone AG <https://greenbone.net>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+// Package Notifications provides a client to communicate with the OpenSight Notification Service [github.com/greenbone/opensight-notification-service]
+package notifications
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+)
+
+const basePath = "/api/notification-service"
+const createNotificationPath = "/notifications"
+
+// Client can be used to send notifications
+type Client struct {
+	httpClient                 *http.Client
+	notificationServiceAddress string
+}
+
+// NewClient returns a new [Client] with the notification service address (host:port) set.
+func NewClient(httpClient *http.Client, notificationServiceAddress string) *Client {
+	return &Client{
+		httpClient:                 httpClient,
+		notificationServiceAddress: notificationServiceAddress,
+	}
+}
+
+// CreateNotification sends a notification to the notification service
+func (c *Client) CreateNotification(notification Notification) error {
+	var notificationSerialized bytes.Buffer
+	err := json.NewEncoder(&notificationSerialized).Encode(notification)
+	if err != nil {
+		return fmt.Errorf("failed to serialize notification object: %w", err)
+	}
+
+	// create request
+	createNotificationEndpoint, err := url.JoinPath(c.notificationServiceAddress, basePath, createNotificationPath)
+	if err != nil {
+		return fmt.Errorf("invalid url '%s': %w", c.notificationServiceAddress, err)
+	}
+	request, err := http.NewRequest(http.MethodPost, createNotificationEndpoint, &notificationSerialized)
+	if err != nil {
+		return fmt.Errorf("failed to build request: %w", err)
+	}
+
+	// execute request
+	response, err := c.httpClient.Do(request)
```

Comment:
how about implementing a retry with exponential backoff for the case that the notification service was not reachable (connection refused)? This cold reduce the risk of losing notifications - and would not require the user of the library to build an appropriate solution.

Reply:
Good idea, I added a package `retryableRequest` which provides a function to execute a request with retries (utilizing some parts of https://github.com/hashicorp/go-retryablehttp). This function is used here. I also added a context to the function signature, so that the client can cancel the retries if needed.